### PR TITLE
use `@ember/test-waiters` to instruct Ember to wait for modal `close`

### PR DESCRIPTION
**Pull request by fivetanley**
### :pushpin: Summary

Makes tests less flakey in CI for consumers of the design system

### :hammer_and_wrench: Detailed description

This PR instructs Ember's test suite to wait for the `<dialog>`'s `close` event.

In one of our product's test suites, we were experiencing some flakiness with tests randomly failing when using modals in tests. This is due to Chrome/other browsers queing the `<dialog>`'s `close` event. From the [HTML Spec][1]:

> Queue an element task on the user interaction task source given the subject
> element to fire an event named close at subject.

This means that the `close` event could fire at the end of the current loop or at the beginning of the next turn of the event loop, or in other words, after our test code has run and waited for side effects. Because Ember is unaware of these events, Ember's test infrastructure isn't waiting for the `close` event to fire when using `await click(selector)`/etc from [@ember/test-helpers][2]. We can't depend on the browser to run these events in a particular order as other events may have been queued first.

In order to help consumers of the design system use this component successfully in their application without flakiness or manual polling logic, we can take advantage of Ember's [@ember/test-waiters library][3], which will [eventually become a standard part of Ember.js][4].

Before we close the `<dialog>` element, we register a `waiter` from `@ember/test-waiters` and then resolve it when the close event fires.

To reduce the impact on production builds, we use `ember-cli-babel`'s [General Purpose Env Flags][5] to strip out the instrumentation code during production builds.

[1]: https://html.spec.whatwg.org/multipage/interactive-elements.html#dom-dialog-close
[2]: https://github.com/emberjs/ember-test-helpers
[3]: https://github.com/emberjs/ember-test-waiters
[4]: https://github.com/emberjs/rfcs/blob/master/text/0581-new-test-waiters.md
[5]: https://github.com/babel/ember-cli-babel#general-purpose-env-flags

### :camera_flash: Screenshots

N/A

### :link: External links

* https://hashicorp.slack.com/archives/C03JP1TCGQM/p1691097709962489

### 👀 Reviewer's checklist:

- [x] +1 Percy if applicable
- [x] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed
- [x] Confirm that A11y tests have been run locally for this component

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

### CODE REVIEW

File: packages/components/addon/components/hds/modal/index.js [hunk 1]
```diff
@@ -8,6 +8,14 @@ import { tracked } from '@glimmer/tracking';
 import { action } from '@ember/object';
 import { assert } from '@ember/debug';
 import { getElementId } from '@hashicorp/design-system-components/utils/hds-get-element-id';
+import { buildWaiter } from '@ember/test-waiters';
+import { DEBUG } from '@glimmer/env';
+
+let waiter;
+
+if (DEBUG) {
```
Comment (fivetanley) [inline packages/components/addon/components/hds/modal/index.js:16]: With Ember-CLI's General Purpose ENV flags, this code will get stripped out in production builds

File: packages/components/addon/components/hds/modal/index.js [hunk 2]
```diff
@@ -8,6 +8,14 @@ import { tracked } from '@glimmer/tracking';
 import { action } from '@ember/object';
 import { assert } from '@ember/debug';
 import { getElementId } from '@hashicorp/design-system-components/utils/hds-get-element-id';
+import { buildWaiter } from '@ember/test-waiters';
+import { DEBUG } from '@glimmer/env';
```
Comment (fivetanley) [inline packages/components/addon/components/hds/modal/index.js:12]: With Ember-CLI's General Purpose ENV flags, this code will get stripped out in production builds

File: packages/components/addon/components/hds/modal/index.js [hunk 3]
```diff
@@ -156,7 +164,18 @@ export default class HdsModalIndexComponent extends Component {
   }
 
   @action
-  onDismiss() {
+  async onDismiss() {
+    // allow ember test helpers to be aware of when the `close` event fires
+    // when using `click` or other helpers from '@ember/test-helpers'
+    if (DEBUG && this.element.open) {
```
Comment (fivetanley) [inline packages/components/addon/components/hds/modal/index.js:170]: With Ember-CLI's General Purpose ENV flags, `DEBUG` will be replaced with `false`, and this code will get stripped out in production builds by the minifier

Comment (fivetanley) [reply packages/components/addon/components/hds/modal/index.js:170]: Checking if the element is open is required, otherwise we'll set up an event listener that will never fire if the modal is already closed.

**Review by jarrodn-au — APPROVED** [review]
Thanks for: the detailed explanation, dealing with such an edge case 😅, reducing test flakiness. Excellent!

File: packages/components/addon/components/hds/modal/index.js (hunk 2 above)
Comment (fivetanley) [reply packages/components/addon/components/hds/modal/index.js:12]: `DEBUG` evaluates to `true` in dev/test builds, but `false` in prod builds

File: packages/components/addon/components/hds/modal/index.js (hunk 1 above)
Comment (didoo) [inline packages/components/addon/components/hds/modal/index.js:16]:
```suggestion
// Notice: this code will get stripped out in production builds (DEBUG evaluates to `true` in dev/test builds, but `false` in prod builds)
if (DEBUG) {
```

File: packages/components/addon/components/hds/modal/index.js (hunk 3 above)
Comment (didoo) [inline packages/components/addon/components/hds/modal/index.js:170]:
```suggestion
    // Notice: this code will get stripped out in production builds (DEBUG evaluates to `true` in dev/test builds, but `false` in prod builds)
    if (DEBUG && this.element.open) {
```

**Review by didoo — APPROVED** [review]
Thanks for this PR ❤️

I have left a couple of suggestion about adding comments (it may not be obvious to other developers, myself included 😀)

**Review by alex-ju — APPROVED** [review]
Thank you for the detailed explanation and for taking the time to debug this! 🙌  Before merging, it would be good if you could add a changelog entry via [`yarn changeset`](https://github.com/hashicorp/design-system/blob/main/README.md?plain=1#L71) (could be as simple as: `Hds::Modal` - fix flaky tests for on `close` event, or however you see fit).
